docs/pipeline.rst

```diff
@@ -143,3 +143,41 @@ DataNode
 --------
 .. autoclass:: nvidia.dali.pipeline.DataNode
    :members:
+
+Pipeline Debug Mode (experimental)
+------------------
+
+Pipeline can be run in debug mode by replacing :meth:`@nvidia.dali.pipeline_def <nvidia.dali.pipeline_def>` decorator with its experimental variant
+``@nvidia.dali.experimental.pipeline_def`` and setting parameter ``debug`` to True. It allows you
+to access and modify data inside of the pipeline execution graph as well as use non-DALI data types
+as inputs for DALI operators.
+
+In this mode outputs of operators are of type ``DataNodeDebug`` which is an equivalent to
+:class:`~nvidia.dali.pipeline.DataNode` in the standard mode. You can perform the same
+operations on objects of type ``DataNodeDebug`` as on ``DataNode``, that includes arithmetic
+operations.
+
+Use ``.get()`` to access data associated with the ``DataNodeDebug`` object during current execution
+of :meth:`Pipeline.run`::
+
+    @nvidia.dali.experimental.pipeline_def(debug=True)
+    def my_pipe():
+        data, _ = fn.readers.file(file_root=images_dir)
+        img = fn.decoders.image(data)
+        print(np.array(img.get()[0]))
+        ...
+
+Use non-DALI data types (e.g. NumPy ndarray, PyTorch Tensor) directly with DALI operators::
+
+    @nvidia.dali.experimental.pipeline_def(batch_size=8, debug=True)
+    def my_pipe():
+        img = [np.random.rand(640, 480, 3) for _ in range(8)]
+        output = fn.flip(img)
+        ...
```

Comment:
Something bugs me here. What error user would get, if he removes `debug=True` in this snippet?

I mean, this is really cool feature (and we certainly want to have it), but potentially might impose a lot of problems for the user when he removes the debug mode. I'm wondering how we could avoid this trouble?

Reply:
Of course there will be cases when you won't be able to transform debug pipeline to a regular one by simply removing `debug=True` and I don't think that we've ever believed it would be otherwise. What we want is to maintain the compatibility the other way, i.e. from regular to debug mode by simply adding `debug=True`. Other features we are adding to the debug mode also make it inconvertible to a regular one (simple `.get()`), you always have to remove those parts of code.